sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+  AZ_PRECONDITION_NOT_NULL(mqtt_user_name);
+  AZ_PRECONDITION(mqtt_user_name_size > 0);
+
+  az_span mqtt_user_name_span
+      = az_span_init((uint8_t*)mqtt_user_name, (int32_t)mqtt_user_name_size);
 
   const az_span* const user_agent = &(client->_internal.options.user_agent);
+  az_span str_registrations
+      = az_span_slice(str_dps_registrations_res, idx_registrations_start, idx_registrations_end);
 
   int32_t required_length = az_span_size(client->_internal.id_scope)
       + az_span_size(str_registrations) + az_span_size(client->_internal.registration_id)
       + az_span_size(provisioning_service_api_version);
   if (az_span_size(*user_agent) > 0)
   {
-    required_length += az_span_size(*user_agent) + 1;
+    required_length += az_span_size(user_agent_version_prefix) + az_span_size(*user_agent);
   }
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_user_name, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_user_name_span, required_length);
 
-  az_span remainder = az_span_copy(mqtt_user_name, client->_internal.id_scope);
+  az_span remainder = az_span_copy(mqtt_user_name_span, client->_internal.id_scope);
   remainder = az_span_copy(remainder, str_registrations);
   remainder = az_span_copy(remainder, client->_internal.registration_id);
   remainder = az_span_copy(remainder, provisioning_service_api_version);
 
   if (az_span_size(*user_agent) > 0)
   {
-    remainder = az_span_copy_u8(remainder, '&');
-    az_span_copy(remainder, *user_agent);
+    remainder = az_span_copy(remainder, user_agent_version_prefix);
+    remainder = az_span_copy(remainder, *user_agent);
   }
 
-  *out_mqtt_user_name = az_span_slice(mqtt_user_name, 0, required_length);
+  remainder = az_span_copy_u8(remainder, 0);
+
+  if (out_mqtt_user_name_length)
+  {
+    *out_mqtt_user_name_length = (size_t)required_length;
+  }
 
   return AZ_OK;
 }
 
 // <registration_id>
-AZ_NODISCARD az_result az_iot_provisioning_client_id_get(
+AZ_NODISCARD az_result az_iot_provisioning_client_get_client_id(
     az_iot_provisioning_client const* client,
-    az_span mqtt_client_id,
-    az_span* out_mqtt_client_id)
+    char* mqtt_client_id,
+    size_t mqtt_client_id_size,
+    size_t* out_mqtt_client_id_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_client_id, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_client_id);
+  AZ_PRECONDITION_NOT_NULL(mqtt_client_id);
+  AZ_PRECONDITION(mqtt_client_id_size > 0);
+
+  az_span mqtt_client_id_span
+      = az_span_init((uint8_t*)mqtt_client_id, (int32_t)mqtt_client_id_size);
 
   int required_length = az_span_size(client->_internal.registration_id);
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_client_id, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_client_id_span, required_length);
 
-  az_span_copy(mqtt_client_id, client->_internal.registration_id);
+  az_span remainder = az_span_copy(mqtt_client_id_span, client->_internal.registration_id);
+  remainder = az_span_copy_u8(remainder, '\0');
```

Comment:
>'\0' [](start = 41, length = 4)

can we either use '\0' or 0 consistently? Line 95 uses 0 and here we are using '\0'

Reply:
if it is a string, we should use '\0' to be clear it's a terminating null